File: src/lib/tauri/route.ts
import { invoke } from '@tauri-apps/api/core';

export interface Route {
  id: string;
  profileId: string;
  hostname: string;
  target: string;
  createdAt: string;
}

export interface CreateRouteInput {
  profileId: string;
  hostname: string;
  target: string;
}

export interface UpdateRouteInput {
  id: string;
  profileId: string;
  hostname: string;
  target: string;
}

export interface RouteError {
  code: string;
  message: string;
}

export interface RouteApi {
  getRoutes(profileId: string): Promise<Route[]>;
  addRoute(input: CreateRouteInput): Promise<Route>;
  updateRoute(input: UpdateRouteInput): Promise<Route>;
  deleteRoute(routeId: string, profileId: string): Promise<void>;
}

export const tauriRouteApi: RouteApi = {
  getRoutes(profileId) {
    return invoke<Route[]>('get_routes', { profileId });
  },
  addRoute(input) {
    return invoke<Route>('add_route', { input });
  },
  updateRoute(input) {
    return invoke<Route>('update_route', { input });
  },
  deleteRoute(routeId, profileId) {
    return invoke<void>('delete_route', { routeId, profileId });
  },
};

File: src-tauri/src/commands/route.rs
use tauri::{AppHandle, Manager};

use crate::domain::route::{CreateRouteInput, Route, UpdateRouteInput};
use crate::error::UiError;
use crate::services::route_service::RouteService;
use crate::store::profile_config_store::ProfileConfigStore;

#[tauri::command]
pub fn get_routes(app: AppHandle, profile_id: String) -> Result<Vec<Route>, UiError> {
    let service = route_service(&app)?;
    service.list_routes(&profile_id).map_err(UiError::from)
}

#[tauri::command]
pub fn add_route(app: AppHandle, input: CreateRouteInput) -> Result<Route, UiError> {
    let service = route_service(&app)?;
    service.add_route(input).map_err(UiError::from)
}

#[tauri::command]
pub fn update_route(app: AppHandle, input: UpdateRouteInput) -> Result<Route, UiError> {
    let service = route_service(&app)?;
    service.update_route(input).map_err(UiError::from)
}

#[tauri::command]
pub fn delete_route(
    app: AppHandle,
    route_id: String,
    profile_id: String,
) -> Result<(), UiError> {
    let service = route_service(&app)?;
    service
        .delete_route(&route_id, &profile_id)
        .map_err(UiError::from)
}

fn route_service(app: &AppHandle) -> Result<RouteService, UiError> {
    let config_dir = app
        .path()
        .app_config_dir()
        .map_err(|_| UiError::from(crate::error::AppError::ConfigReadFailed))?;
    let config_store = ProfileConfigStore::new(config_dir.join("profile.json"));
    Ok(RouteService::new(config_store))
}
